io.home-assistant/index.js

```diff
@@ -0,0 +1,279 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of io.home-assistant
+//
+// Copyright 2019 Giovanni Campagna <gcampagn@cs.stanford.edu>
+//
+// See LICENSE for details
+"use strict";
+
+const Tp = require('thingpedia');
+const WebSocket = require('ws');
+const HomeAssistant = require('home-assistant-js-websocket');
+
+// FIXME make configurable
+const HASS_URL = 'http://hassio.local:8123';
+
+class HomeAssistantDevice extends Tp.BaseDevice {
+    constructor(engine, state, master, entityId) {
+        super(engine, state);
+
+        this.master = master;
+        this._entityId = entityId;
+        this.uniqueId = master.uniqueId + '/' + entityId;
+        this.name = this.state.friendly_name;
+    }
+
+    updateState(state) {
+        super.updateState(state);
+        this.name = this.state.friendly_name;
+    }
+
+    _callService(domain, service, data = {}) {
+        data.entity_id = this._entityId;
+        return HomeAssistant.callService(this.master.connection, domain, service, data);
+    }
+}
+
+class HomeAssistantLightbulbDevice extends HomeAssistantDevice {
+    constructor(engine, state, master, entityId) {
+        super(engine, state, master, entityId);
+
+        this.uniqueId = master.uniqueId + '/' + entityId;
+    }
+
+    async do_set_power({ power }) {
+        if (power === 'on')
+            await this._callService('light', 'turn_on');
+        else
+            await this._callService('light', 'turn_off');
+    }
+
+    async do_alert_long() {
+        await this._callService('light', 'turn_on', { flash: 'long' });
+    }
+
+    async do_color_loop() {
+        await this._callService('light', 'turn_on', { effect: 'colorloop' });
+    }
+}
+
+class HomeAssistantDeviceSet extends Tp.Helpers.ObjectSet.Base {
+    constructor(master) {
+        super();
+        this.master = master;
+        this._devices = new Map;
+
+        this._unsubscribe = null;
+    }
+
+    _maybeAddEntity(entityId, entityData) {
+        const existing = this._devices.get(entityId);
+        if (existing) {
+            // add "kind" property that Tp expects
+            entityData.kind = existing.state.kind;
+            existing.updateState(entityData);
+            return;
+        }
+
+        const [type,] = entityId.split('.');
```

Comment:
The Home Assistant term we use for the first part is `domain`. `split` also takes an optional limit parameter that you could use.